### PR TITLE
[processor/geoip] Add GeoIP providers configuration and maxmind factory

### PR DESCRIPTION
**Pull request by rogercoll**
**Description:**  Define and parse a configuration for the geo IP providers section. In addition, the Maxmind GeoIPProvider implementation is included in the providers factories. 

Example configuration:

```yaml
processors:
  geoip:
    providers:
      maxmind:
        database: "/tmp"
```

Implementation details:
- Custom Unmarshal implementation for the processor's component, this is needed to dynamically load each provider's configuration: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33268/files#diff-aed2c6fd774ef54a3039647190c67e28bd0fc67e008fdd5630b6201c550bd00aR46 . The approach is very similar to how the [hostmetrics receiver loads](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/hostmetricsreceiver/config.go#L44) its scraper configuration.
    - A new factory for the providers is included in order to retrieve their default configuration and get the actual provider: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33268/files#diff-2fbb171efac07bbf07c1bcb67ae981eb481e56491add51b6a137fd2c17eec9dcR27

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/33269

**Testing:** 

- Unit tests for the configuration unmarshall + factories
   - A base mock structure is used through all the test files: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33268/files#diff-28f4a173f1f4b5ccd3cf4c9f7f7b6bf864ef1567a28291322d7e94a9f63243aeR26
- Integration test to verify the behaviour of the processor + maxmind provider
  - The generation of the Maxmind databases has been moved to a [public internal package](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33268/files#diff-83fc0ce7aa1f0495b4f4e5d5aabc2918162fec31ad323cc417b3f8c8eb5a00bcR14) as being used for the unit and integration tests. Should we upload the assembled database files instead? 

**Documentation:** TODO

### CODE REVIEW

File: .chloggen/geoipprocessor_provider.yaml [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: 'enhancement'
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: geoipprocessor
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Add providers configuration and metadata retrival abstraction
```
Comment (andrzej-stencel) [inline .chloggen/geoipprocessor_provider.yaml:10]:
```suggestion
note: Add providers configuration and metadata retrieval abstraction
```

File: processor/geoipprocessor/factory.go [hunk 2]
```diff
@@ -25,25 +28,78 @@ var (
 	}
 )
 
+var providerFactories = map[string]provider.GeoIPProviderFactory{
+	maxmind.TypeStr: &maxmind.Factory{},
```
Comment (rogercoll) [inline processor/geoipprocessor/factory.go:32]: Adding maxmind factory to default available providers

File: processor/geoipprocessor/README.md [hunk 3]
```diff
@@ -11,8 +11,42 @@
...
+```
+
+## Configuration
+
+The following settings must be configured:
+
+- `providers`: A map containing geographical location information providers. These providers are used to search for the geographical location attributes associated with an IP. Supported providers:
+  - [maxmind](./internal/provider/maxmindprovider/README.md)
+
+#### Examples
```
Comment (andrzej-stencel) [inline processor/geoipprocessor/README.md:43]: Why four `#`s?

```suggestion
## Examples
```

**Comment by rogercoll** [comment]
@michalpristas @andrzej-stencel Just moved this PR to be ready for review. Please let me know your thoughts/concerns. The PR is quite large as I included the MaxMind provider in the providers' factories. I am open to moving that part to another PR if preferred.

**Review by andrzej-stencel — APPROVED** [review]

**Pull request merged**